clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java

```diff
@@ -168,6 +171,8 @@ private void processApplicationEvents() {
                 if (event instanceof CompletableEvent)
                     applicationEventReaper.add((CompletableEvent<?>) event);
 
+                List<CompletableEvent<?>> completableEvents = reapExpiredApplicationEvents(currentTimeMs);
```

Comment:
Do we need to reap the events on every loop?

Reply:
It seems like we should remove the reaping from this method entirely and have `runOnce()` call `maybeFailOnMetadataError()` directly.

Reply:
agree we shouldn't involve the reaper before processing events (it could expire events that haven't had the chance to run once, ie. timeout 0). 

Alternatively, we have the completable `event` in hand, so what about something like `maybeFailOnMetadataError(List.of(event))` right after adding it to the reaper on ln 170? (we only care about CompletableEvents anyways), would that work @m1a2st ?